### PR TITLE
Walkthrough 1a

### PR DESCRIPTION
**Pull request by witmicko**

### CODE REVIEW

File: src/pages/landing/landingPage.js [hunk 1]
```diff
@@ -46,7 +47,8 @@ LandingPage.defaultProps = {
 };
 
 const mapDispatchToProps = dispatch => ({
-  manageWalkthroughServices: () => manageMiddlewareServices(dispatch),
+  // manageWalkthroughServices: () => manageMiddlewareServices(dispatch),
+  manageWalkthroughServices: () => prepareWalkthroughNamespace(dispatch, walkthroughs.oneA),
```
Comment (aidenkeating) [inline src/pages/landing/landingPage.js]: I don't think we want to prepare the walkthrough namespaces here. On the landing page we want to provision the ServiceInstance namespace. We probably want to provision the Walkthrough namespace when a user enters the walkthrough itself.

File: src/services/walktroughServices.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+class NodeMessagingAppEnmassSiTransform {
+  isTransformable(siInfo) {
+    return siInfo.name === 'nodejs-messaging-work-queue-frontend';
+  }
+
+  transform(siInfo) {
+    const enmassCreds = {
+      MESSAGING_SERVICE_USER: 'username',
+      MESSAGING_SERVICE_PASSWORD: 'password'
+    };
```
Comment (aidenkeating) [inline src/services/walktroughServices.js]: We'll probably need to set the internal service address here too to override the default URL.

File: src/pages/landing/landingPage.js (hunk 1 above)
Comment (witmicko) [reply src/pages/landing/landingPage.js]: +1, but create that after the user clicks on "Getting started" we don't want to setup the workspace if the user just wants to read the description before choosing the one walkthrough to do.

File: src/services/walktroughServices.js (hunk 2 above)
Comment (JameelB) [reply src/services/walktroughServices.js]: @aidenkeating Is that this URL here? `amqp://messaging.enmasse-eval.svc:5672?amqp.saslMechanisms=PLAIN`

File: src/services/walktroughServices.js [hunk 3]
```diff
@@ -0,0 +1,109 @@
+import { create, watch, currentUser, OpenShiftWatchEvents } from './openshiftServices';
+import {
+  cleanUsername,
+  findOpenshiftResource,
+  buildServiceInstanceResourceObj,
+  findOrCreateOpenshiftResource
+} from './middlewareServices';
```
Comment (JameelB) [inline src/services/walktroughServices.js]: Can we move `cleanUsername`, `findOpenshiftResource` and `findOrCreateOpenshiftResource` functions into commons since these can be reused here. I find it weird to be importing from `middlewareServices` here. wdyt?

I think the `buildServiceInstanceResourceObj` can be imported from `serviceInstanceHelpers`.

File: src/services/walktroughServices.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+const walkthroughs = {
+  one: {
+    namespaceSuffix: 'walkthrough-one',
+    services: ['spring-boot-rest-http-crud', 'nodejs-messaging-work-queue-frontend'],
+    transforms: [new DefaultServiceInstanceTransform()]
+  },
+  oneA: {
+    namespaceSuffix: 'walkthrough-one-a',
+    services: ['spring-boot-rest-http-crud', 'nodejs-messaging-work-queue-frontend'],
+    transforms: [new NodeMessagingAppEnmassSiTransform(), new DefaultServiceInstanceTransform()]
```
Comment (JameelB) [inline src/services/walktroughServices.js]: Is there a reason why we're including `DefaultServiceInstanceTransform` here?

File: src/services/walktroughServices.js [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+  }
+};
+
+const prepareWalkthroughNamespace = (dispatch, walkthrough) => {
+  currentUser().then(user => {
+    const userNamespace = `${cleanUsername(user.username)}-${walkthrough.namespaceSuffix}`;
+    const namespaceObj = osKinds.buildNamespaceObj(userNamespace);
+    findOpenshiftResource(osKinds.projects, namespaceObj)
```
Comment (JameelB) [inline src/services/walktroughServices.js]: I think the function `findOrCreateOpenshiftResource` does the check below for you.

File: src/services/walktroughServices.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
+import { create, watch, currentUser, OpenShiftWatchEvents } from './openshiftServices';
```
Comment (JameelB) [inline src/services/walktroughServices.js]: The filename for `walktroughServices.js` is missing an h => `walkthroughServices.js`

File: src/services/walktroughServices.js (hunk 4 above)
Comment (witmicko) [reply src/services/walktroughServices.js]: hey, it transforms the spboot app, at the moment only node app needs parameters
I was going to include defaulttransform as a fallback, so we dont have to always add it as last, but decided against it as it might cause confusion.

File: src/services/walkthroughServices.js [hunk 7]
```diff
@@ -14,12 +14,14 @@ class NodeMessagingAppEnmassSiTransform {
   }
 
   transform(siInfo) {
-    const enmassCreds = {
+    const placeholderParameters = {
       MESSAGING_SERVICE_USER: 'username',
-      MESSAGING_SERVICE_PASSWORD: 'password'
+      MESSAGING_SERVICE_PASSWORD: 'password',
+      MESSAGING_SERVICE_HOST: 'messaging.enmasse-eval.svc',
```
[the remark this answers is not in the document]
Comment (aidenkeating) [reply src/services/walkthroughServices.js]: Yep, all these will need to be retrieved from EnMasse as they're all namespace specific. We can build this up on the client side if we want though as we know the name of the namespace (since we create it) and we know the service will be called messaging. So basically replace `enmasse-eval` with the known namespace name.

File: src/pages/landing/landingPage.js (hunk 1 above)
Comment (aidenkeating) [reply src/pages/landing/landingPage.js]: Not sure if you've resolved this already. But the approach I took is to start creating the services when the user enters the task pages instead of the tutorial pages. This ensures they've entered the steps of the tutorial.

File: src/services/walkthroughServices.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (witmicko) [reply src/services/walkthroughServices.js]: yup, I was waiting for that to be ready, hence `placeholderParams` as a name

[the remark this answers is not in the document]
Comment (JameelB) [reply src/services/walkthroughServices.js]: I've added the url now as well: https://github.com/integr8ly/tutorial-web-app/pull/60/files#diff-ed7417901bdafe31ec92bf19d0ea8a42R53

**Comment by dlabaj** [comment]
@witmicko Looks like there is merge conflicts with this PR now.

File: src/pages/landing/landingPage.js (hunk 1 above)
Comment (dlabaj) [reply src/pages/landing/landingPage.js]: Agree with @aidenkeating we should prepare the walkthrough when the user enters the tutorial page.

File: src/common/openshiftKinds.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
+const namespaceRequestDef = {
```
Comment (dlabaj) [inline src/common/openshiftKinds.js]: Not sure what is meant by Kinds?  Any thought on a different name ... maybe openshiftHelpers.js?

File: src/services/middlewareServices.js [hunk 9]
```diff
@@ -1,16 +1,13 @@
...
-  namespaceRequestDef,
-  namespaceDef,
-  namespaceRequestResource,
-  statefulSetDef,
-  routeDef,
-  secretDef
+  findOpenshiftResource,
+  findOrCreateOpenshiftResource
 } from '../common/openshiftHelpers';
+import * as osKinds from '../common/openshiftKinds';
```
Comment (dlabaj) [inline src/services/middlewareServices.js]: Can we just import what is needed instead of all the exports?  What do you think?

File: src/common/openshiftKinds.js (hunk 8 above)
Comment (witmicko) [reply src/common/openshiftKinds.js]: I can move it to helpers, but wanted to have k8s/openshift types in a separate file, as they are mostly simple objects

**Comment by witmicko** [comment]
@dlabaj this is ready for review if you have some time.

File: src/common/serviceInstanceHelpers.js [hunk 10]
```diff
@@ -13,7 +17,7 @@ class DefaultServiceInstanceTransform {
       },
       spec: {
         clusterServiceClassExternalName: siInfo.name,
-        clusterServicePlanExternalName: `default-${siInfo.name}`
+        clusterServicePlanExternalName: `default`
```
Comment (JameelB) [inline src/common/serviceInstanceHelpers.js]: @witmicko Can we revert the change to this? This is causing the provison of che, launcher and fuse to fail because the plan name we currently have to the broker is in this format: `default-${siInfo.name}`

**Comment by witmicko** [comment]
@JameelB @aidenkeating addressed comments and squashed, that had dismissed your reviews
